pkg/server/container_stats_list.go

```diff
@@ -17,14 +17,157 @@ limitations under the License.
 package server
 
 import (
-	"errors"
-
+	"fmt"
+	"github.com/golang/glog"
 	"golang.org/x/net/context"
 
+	"github.com/containerd/cgroups"
+	tasks "github.com/containerd/containerd/api/services/tasks/v1"
+	"github.com/containerd/containerd/api/types"
+	"github.com/containerd/typeurl"
 	"k8s.io/kubernetes/pkg/kubelet/apis/cri/v1alpha1/runtime"
 )
 
 // ListContainerStats returns stats of all running containers.
-func (c *criContainerdService) ListContainerStats(ctx context.Context, in *runtime.ListContainerStatsRequest) (*runtime.ListContainerStatsResponse, error) {
-	return nil, errors.New("not implemented")
+func (c *criContainerdService) ListContainerStats(
+	ctx context.Context,
+	in *runtime.ListContainerStatsRequest,
+) (*runtime.ListContainerStatsResponse, error) {
+	request, candidateContainers, err := c.buildTaskMetricsRequest(in)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build metrics request: %v", err)
+	}
+	resp, err := c.taskService.Metrics(context.Background(), &request)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch metrics for tasks: %v", err)
+	}
+	criStats, err := c.toCRIContainerStats(resp.Metrics, candidateContainers)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert to cri containerd stats format: %v", err)
+	}
+	return criStats, nil
+}
+
+func (c *criContainerdService) toCRIContainerStats(
+	stats []*types.Metric,
+	candidateContainers map[string]bool,
+) (*runtime.ListContainerStatsResponse, error) {
+	containerStats := new(runtime.ListContainerStatsResponse)
+	for _, stat := range stats {
+		var cs runtime.ContainerStats
+		if err := c.getContainerMetrics(stat.ID, stat, &cs); err != nil {
+			glog.Errorf("failed to decode container metrics: %v", err)
+			continue
+		}
+		delete(candidateContainers, stat.ID)
+		containerStats.Stats = append(containerStats.Stats, &cs)
+	}
+	// If there is a transient state where containers are dead at the time of query
```

Comment:
It's not a transient state. For any dead container, this will happen until we remove the container.